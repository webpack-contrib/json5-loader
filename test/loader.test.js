import JSON5 from 'json5';

import loader from '../src';

describe('loader', () => {
  it('should export the loader', (done) => {
    expect(loader).toBeInstanceOf(Function);

    done();
  });

  it('should handle valid JSON5', (done) => {
    const json5 = `{
  // comments
  unquoted: 'and you can quote me on that',
  singleQuotes: 'I can use "double quotes" here',
  lineBreaks: "Look, Mom! \\
No \\\\n's!",
  hexadecimal: 0xdecaf,
  leadingDecimalPoint: .8675309, andTrailing: 8675309.,
  positiveSign: +1,
  trailingComma: 'in objects', andIn: ['arrays',],
  "backwardsCompatible": "with JSON",
}`;
    const emitError = jest.fn();
    const content = loader.call(
      {
        emitError,
      },
      json5
    );

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should invalid characters', (done) => {
    const json5 = `{foo: '\u2028\u2029'}`;
    const emitError = jest.fn();
    const content = loader.call(
      {
        emitError,
      },
      json5
    );

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should preserve Infinity', (done) => {
    const json5 = '{to : Infinity}';
    const content = loader.call({}, json5);

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should preserve -Infinity', (done) => {
    const json5 = '{to : -Infinity}';
    const content = loader.call({}, json5);

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should preserve null', (done) => {
    const json5 = '{null : null}';
    const content = loader.call({}, json5);

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should preserve NaN', (done) => {
    const json5 = '{nan : NaN}';
    const content = loader.call({}, '{nan : NaN}');

    expect(content).toMatchSnapshot();

    // eslint-disable-next-line no-eval
    const js = eval(content);

    expect(JSON5.stringify(js)).toEqual(JSON5.stringify(JSON5.parse(json5)));

    done();
  });

  it('should handle invalid JSON5', (done) => {
    const brokenJson5 = '{broken: json5}';
    const emitError = jest.fn();

    loader.call(
      {
        emitError,
      },
      brokenJson5
    );

    expect(emitError).toHaveBeenCalledWith(expect.any(SyntaxError));

    done();
  });
});
