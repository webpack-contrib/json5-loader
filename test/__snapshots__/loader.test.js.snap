// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should handle valid JSON5 1`] = `"module.exports = {unquoted:'and you can quote me on that',singleQuotes:'I can use \\"double quotes\\" here',lineBreaks:\\"Look, Mom! No \\\\\\\\n's!\\",hexadecimal:912559,leadingDecimalPoint:0.8675309,andTrailing:8675309,positiveSign:1,trailingComma:'in objects',andIn:['arrays'],backwardsCompatible:'with JSON'}"`;

exports[`loader should invalid characters 1`] = `"module.exports = {foo:'\\\\u2028\\\\u2029'}"`;

exports[`loader should preserve -Infinity 1`] = `"module.exports = {to:-Infinity}"`;

exports[`loader should preserve Infinity 1`] = `"module.exports = {to:Infinity}"`;

exports[`loader should preserve NaN 1`] = `"module.exports = {nan:NaN}"`;

exports[`loader should preserve null 1`] = `"module.exports = {null:null}"`;
